src/charm.py

```diff
@@ -561,29 +583,102 @@ def _on_mongod_pebble_ready(self, event) -> None:
+
+        # Add Pebble config layers missing or modified
+        for layer_name, definition in new_layers.items():
+            for service_name, service in definition.services.items():
+                if current_layers.services.get(service_name) != service:
+                    modified = True
+                    logger.debug(f"Adding layer {service_name}.")
+                    container.add_layer(layer_name, definition, combine=True)
 
+        # We'll always have a logrotate configuration at this point.
         container.exec(["chmod", "644", "/etc/logrotate.d/mongodb"])
 
-        # Restart changed services and start startup-enabled services.
-        container.replan()
+        if modified:
+            container.replan()
+
+    def _configure_container(self, container: Container):
+        """Configure MongoDB pebble layer specification."""
+        if not container.can_connect():
+            logger.debug("mongod container is not ready yet.")
+            raise ContainerNotReadyError
+
+        # We need to check that the storages are attached before starting the services.
+        # pebble-ready is not guaranteed to run after storage-attached so this check allows
+        # to ensure that the storages are attached before the pebble-ready hook is run.
+        if any(not storage for storage in self.model.storages.values()):
+            logger.debug("Storages are not attached yet")
+            raise ContainerNotReadyError
+
+        try:
+            self.__filesystem_handler(container)
+        except FailedToElectNewPrimaryError as err:
+            raise ContainerNotReadyError from err
+
+        self.__configure_layers(container)
 
         # when a network cuts and the pod restarts - reconnect to the exporter
         try:
             self._connect_mongodb_exporter()
             self._connect_pbm_agent()
         except MissingSecretError as e:
             logger.error("Cannot connect mongodb exporter: %r", e)
+            raise ContainerNotReadyError
+
+    # BEGIN: charm events
+    def _on_upgrade(self, event: UpgradeCharmEvent):
+        """Upgrade event handler.
+
+        During an upgrade event, it will set the version in all relations,
+        replan the container and process the upgrade statuses. If the upgrade
+        is compatible, it will end up emitting a post upgrade event that
+        verifies the health of the cluster.
+        """
+        if self.unit.is_leader():
+            self.version_checker.set_version_across_all_relations()
+
+        container = self.unit.get_container(Config.CONTAINER_NAME)
+
+        # Just run the configure layers steps on the container and defer if it fails.
+        try:
+            self._configure_container(container)
+        except ContainerNotReadyError:
+            self.status.set_and_share_status(Config.Status.UNHEALTHY_UPGRADE)
             event.defer()
             return
 
+        self.status.set_and_share_status(ActiveStatus())
```

Comment:
If we don't do that, the upgrade code status setting does nothing.
It needs to have at least `ActiveStatus` or a specific `BlockedStatus`.
In the VM case, it's not an issue because the charm is not in maintenance mode during upgrade, but this is the case during a k8s upgrade (and it makes all tests afterward based on status fail).
I think it's safe considering that right after, the status will be managed by `_reconcile_upgrade` or the `post_app_upgrade` event.
I can do something better I think using `self.status.process_statuses`, I'll see how it works!

Reply:
I did something better :)